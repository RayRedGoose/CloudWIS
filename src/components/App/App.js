import React, { Component } from 'react'
import './App.scss'
import utility from '../../_utility/utility'
import { getCurrentWeatherData, getWeekWeatherData } from '../../apiCalls/apiCalls'
import Board from '../Board/Board'
import LoadingImage from '../LoadingImage/LoadingImage'
import AppHeader from '../AppHeader/AppHeader'

class App extends Component {
  constructor() {
    super()
    this.state = {
      data: {
        currentCity: 'Denver',
        searchHistory: ['Denver', 'Phoenix', 'Houston'],
        weatherData: {
          today: null,
          week: null
        }
      },
      sys: {
        error: null,
        isLoaded: false
      }
    }
  }

  componentDidMount = () => {
    this.getDataFromLocalStorage()
      .then((currentCity) => this.fetchWeatherData(currentCity))
  }

  async getDataFromLocalStorage() {
    const currentCity = (localStorage.length)
      ? localStorage.getItem('currentCity')
      : 'Denver'
    const searchHistory = (localStorage.length)
      ? localStorage.getItem('searchHistory').split(',')
      : ['Denver', 'Phoenix', 'Houston']

    this.setState({
      ...this.state,
      data: {
        currentCity: currentCity,
        searchHistory: searchHistory,
        weatherData: {
          today: null,
          week: null
        }
      }
    })
    return currentCity
  }

  fetchWeatherData = (city) => {
    Promise.all([
      getCurrentWeatherData(city)
        .then(data => this.setDayWeatherInfo(data))
        .catch(error => this.setError(error)),

      getWeekWeatherData(city)
        .then(data => this.setWeekWeatherInfo(data))
    ]).then(() => this.setLoaded(true))
      .catch(error => this.setError(error))
  }

  setDayWeatherInfo = (data) => {
    this.setState({
      ...this.state,
      data: {
        ...this.state.data,
        weatherData: {
          ...this.state.data.weatherData,
          today: utility.updateTodayData(data),
        }
      }
    })
  }

  setWeekWeatherInfo = (data) => {
    this.setState({
      ...this.state,
      data: {
        ...this.state.data,
        weatherData: {
          ...this.state.data.weatherData,
          week: utility.updateDaysData(data)
        }
      }
    })
  }

  setError = (error) => {
    this.setState({
      data: {...this.state.data},
      sys: {
        ...this.state.sys,
        error: error.message
      }
    })
  }

  setLoaded = (state) => {
    this.setState({
      data: {...this.state.data},
      sys: {
        ...this.state.sys,
        isLoaded: state
      }
    })
  }

  changeCity = (city) => {
    const cities =
      (!this.state.data.searchHistory.includes(city))
        ? [...this.state.data.searchHistory, city]
        : [...this.state.data.searchHistory]

    this.setState({
      data: {
        currentCity: city,
        searchHistory: cities,
        weatherData: {
          today: null,
          week: null
        }
      },
      sys: {
        error: null,
        isLoaded: false
      }
    })

    localStorage.setItem('currentCity', city)
    localStorage.setItem('searchHistory', cities)
  }

  render() {
    return (
      <main className="app">
        <AppHeader />
        {
          (this.state.sys.isLoaded)
            ? <Board
              {...this.state.data}
              changeCity={this.changeCity}
              fetchWeatherData={this.fetchWeatherData}/>
            : <LoadingImage />
        }
      </main>
    )

  }
}

export default App
