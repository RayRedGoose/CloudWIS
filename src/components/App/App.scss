.app {
  padding: 5vh 7.5vw;
  &>header {
    position: absolute;
    top: 0; right: 0;
    h1 {
      color: #017E91;
      display: flex;
      align-items: center;
      padding: 20px;
      &:hover {
        color: #fff;
      }
      &:before {
        content: '';
        width: 36px; height: 36px;
        margin-right: 10px;
        background-image: url('../../images/logo.svg');
      }
      a {
        color: inherit;
        text-decoration: none;
      }
    }
  }
}
