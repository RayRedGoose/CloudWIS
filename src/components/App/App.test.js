import React from 'react'
import { shallow } from 'enzyme'
import App from './App'

describe('App', () => {
  let app

  beforeEach(() => {
    app = shallow(<App />)
  })

  it('should match the snapshot with no data fetched', () => {
    expect(app).toMatchSnapshot()
  })

  it('should match the snapshot with all data fetched', () => {
    app.setState({ sys: {
      isLoaded: true
    }})

    expect(app).toMatchSnapshot()
  })

  describe('change city method', () => {
    it('should update current city when changeCity is called', () => {
      // Expectation
      expect(app.state('data').currentCity).toEqual('Denver')

      // Execution
      app.instance().changeCity('Phoenix')

      // Expectation after
      expect(app.state('data').currentCity).toEqual('Phoenix')
    })

    it('should add new city to search history when changeCity is called', () => {
      const cities = [ 'Denver', 'Phoenix','Houston' ]
      // Expectation
      expect(app.state('data').searchHistory).toEqual(cities)

      // Execution
      app.instance().changeCity('Boise')

      // Expectation after
      expect(app.state('data').searchHistory).toEqual([
        ...cities,
        'Boise'
      ])
    })

    it('should keep search history the same if new city is in history already when changeCity is called', () => {
      const cities = [ 'Denver', 'Phoenix','Houston' ]
      // Expectation
      expect(app.state('data').searchHistory).toEqual(cities)

      // Execution
      app.instance().changeCity('Phoenix')

      // Expectation after
      expect(app.state('data').searchHistory).toEqual(cities)
    })

  })

  describe('setDayWeatherInfo method', () => {
    let mockData


    beforeEach(() => {
      mockData = {
        weather: [{main: 'Snow'}],
        main: {temp: 300, humidity: 87},
        wind: {speed: 1},
        visibility: 16093.44
      }
    })

    it('should add weather description when setDayWeatherInfo is called', () => {
      // Expectation
      expect(app.state('data').weatherData.today)
        .toEqual(null)

      // Execution
      app.instance().setDayWeatherInfo(mockData)

      // Expectation after
      expect(app.state('data').weatherData.today.desc)
        .toEqual('snow')
    })

    it('should add temperature when setDayWeatherInfo is called', () => {
      // Expectation
      expect(app.state('data').weatherData.today)
        .toEqual(null)

      // Execution
      app.instance().setDayWeatherInfo(mockData)

      // Expectation after
      expect(app.state('data').weatherData.today.temp)
        .toEqual(80)
    })

    it('should add wind speed when setDayWeatherInfo is called', () => {
      // Expectation
      expect(app.state('data').weatherData.today)
        .toEqual(null)

      // Execution
      app.instance().setDayWeatherInfo(mockData)

      // Expectation after
      expect(app.state('data').weatherData.today.wind)
        .toEqual(2)
    })

    it('should add humidity when setDayWeatherInfo is called', () => {
      // Expectation
      expect(app.state('data').weatherData.today)
        .toEqual(null)

      // Execution
      app.instance().setDayWeatherInfo(mockData)

      // Expectation after
      expect(app.state('data').weatherData.today.humidity)
        .toEqual(87)
    })

    it('should add visibility when setDayWeatherInfo is called', () => {
      // Expectation
      expect(app.state('data').weatherData.today)
        .toEqual(null)

      // Execution
      app.instance().setDayWeatherInfo(mockData)

      // Expectation after
      expect(app.state('data').weatherData.today.visibility)
        .toEqual(10)
    })
  })

  describe('setWeekWeatherInfo method', () => {
    let mockData

    beforeEach(() => {
      mockData = {
        list: [{
          dt_txt: '2019-12-02 15:00',
          main: {temp: 300},
          weather: [{main: 'Snow'}]
        }, {
          'dt_txt': '2019-12-03 15:00',
          main: {temp: 270},
          weather: [{main: 'Rain'}]
        }]
      }
    })

    it('should add short week weather info when setWeekWeatherInfo is called', () => {
      // Expectation
      expect(app.state('data').weatherData.week)
        .toEqual(null)

      // Execution
      app.instance().setWeekWeatherInfo(mockData)

      // Expectation after
      expect(app.state('data').weatherData.week)
        .toEqual({
          'Dec,02': '80.33-snow',
          'Dec,03': '26.33-rain'
        })
    })
  })

  it('should update loaded status when all data was fetched is called', () => {
    // Expectation
    expect(app.state('sys').isLoaded).toEqual(false)

    // Execution
    app.instance().setLoaded(true)

    // Expectation after
    expect(app.state('sys').isLoaded).toEqual(true)
  })

  it('should add error information when setError is called', () => {
    // Setup
    const error = { message: 'There is an error' }

    // Expectation
    expect(app.state('sys').error).toEqual(null)

    // Execution
    app.instance().setError(error)

    // Expectation after
    expect(app.state('sys').error).toEqual('There is an error')
  })

})
