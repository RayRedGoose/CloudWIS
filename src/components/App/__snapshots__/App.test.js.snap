// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all data fetched 1`] = `
<main
  className="app"
>
  <AppHeader />
  <Board
    changeCity={[Function]}
    currentCity="Denver"
    fetchWeatherData={[Function]}
    searchHistory={
      Array [
        "Denver",
        "Phoenix",
        "Houston",
      ]
    }
    weatherData={
      Object {
        "today": null,
        "week": null,
      }
    }
  />
</main>
`;

exports[`App should match the snapshot with no data fetched 1`] = `
<main
  className="app"
>
  <AppHeader />
  <LoadingImage />
</main>
`;
